### PR TITLE
Add additional track pieces to LIM Launched Coaster and the Looping Coaster

### PR DESCRIPTION
**Pull request by X123M3-256**
This PR adds additional track pieces to the LIM/Looping coaster track. The following track pieces are made available by default on the LIM launched coaster:

* Large flat to steep
* Level banked turn to unbanked slope
* Small half loops
* Medium half loops
* Large half loops
* Large corkscrews
* Barrel rolls

The following track pieces are added to the looping coaster:

* Large steep to flat
* Level banked turn to unbanked slope

The following track pieces are also added, but not made available by default:

* Small zero G roll
* Large zero G roll
* Small flat to steep

### CODE REVIEW

**Comment by karst** [comment]
ImageId has recently been refactored in #18126, might be that the build is failing because of that.

File: src/openrct2/ride/coaster/meta/LIMLaunchedRollerCoaster.h [hunk 1]
```diff
@@ -19,8 +19,8 @@ constexpr const RideTypeDescriptor LIMLaunchedRollerCoasterRTD =
...
     SET_FIELD(Category, RIDE_CATEGORY_ROLLERCOASTER),
-    SET_FIELD(EnabledTrackPieces, {TRACK_STRAIGHT, TRACK_STATION_END, TRACK_FLAT_ROLL_BANKING, TRACK_VERTICAL_LOOP, TRACK_SLOPE, TRACK_SLOPE_STEEP_UP, TRACK_SLOPE_STEEP_DOWN, TRACK_SLOPE_CURVE, TRACK_SLOPE_CURVE_STEEP, TRACK_S_BEND, TRACK_CURVE_SMALL, TRACK_CURVE, TRACK_TWIST, TRACK_CORKSCREW, TRACK_HELIX_SMALL, TRACK_BRAKES, TRACK_ON_RIDE_PHOTO, TRACK_SLOPE_VERTICAL, TRACK_BLOCK_BRAKES, TRACK_SLOPE_ROLL_BANKING, TRACK_CURVE_VERTICAL, TRACK_QUARTER_LOOP}),
-    SET_FIELD(ExtraTrackPieces, {}),
+    SET_FIELD(EnabledTrackPieces, {TRACK_STRAIGHT, TRACK_STATION_END, TRACK_FLAT_ROLL_BANKING, TRACK_VERTICAL_LOOP, TRACK_SLOPE, TRACK_SLOPE_STEEP_UP, TRACK_SLOPE_STEEP_DOWN, TRACK_SLOPE_CURVE, TRACK_SLOPE_CURVE_STEEP, TRACK_S_BEND, TRACK_CURVE_SMALL, TRACK_CURVE, TRACK_TWIST, TRACK_CORKSCREW, TRACK_HELIX_SMALL, TRACK_BRAKES, TRACK_ON_RIDE_PHOTO, TRACK_SLOPE_VERTICAL, TRACK_BLOCK_BRAKES, TRACK_SLOPE_ROLL_BANKING, TRACK_CURVE_VERTICAL, TRACK_QUARTER_LOOP,  TRACK_BARREL_ROLL, TRACK_SLOPE_CURVE_BANKED, TRACK_SLOPE_STEEP_LONG, TRACK_HALF_LOOP, TRACK_HALF_LOOP_MEDIUM, TRACK_HALF_LOOP_LARGE, TRACK_CORKSCREW_LARGE}),
```
Comment (Generalcamo) [inline src/openrct2/ride/coaster/meta/LIMLaunchedRollerCoaster.h:22]: The addition of the barrel roll to the LIM coaster should be noted in the changelog

File: src/openrct2/ride/coaster/meta/LoopingRollerCoaster.h [hunk 2]
```diff
@@ -19,8 +19,8 @@ constexpr const RideTypeDescriptor LoopingRollerCoasterRTD =
...
-    SET_FIELD(ExtraTrackPieces, {TRACK_SLOPE_VERTICAL, TRACK_TWIST, TRACK_CORKSCREW, TRACK_CURVE_VERTICAL, TRACK_QUARTER_LOOP}),
+    SET_FIELD(EnabledTrackPieces, {TRACK_STRAIGHT, TRACK_STATION_END, TRACK_LIFT_HILL, TRACK_FLAT_ROLL_BANKING, TRACK_VERTICAL_LOOP, TRACK_SLOPE, TRACK_SLOPE_STEEP_UP, TRACK_SLOPE_STEEP_DOWN, TRACK_SLOPE_CURVE, TRACK_SLOPE_CURVE_STEEP, TRACK_S_BEND, TRACK_CURVE_SMALL, TRACK_CURVE, TRACK_HELIX_SMALL, TRACK_BRAKES, TRACK_ON_RIDE_PHOTO, TRACK_BLOCK_BRAKES, TRACK_SLOPE_ROLL_BANKING, TRACK_BOOSTER, TRACK_SLOPE_CURVE_BANKED, TRACK_SLOPE_STEEP_LONG}),
+    SET_FIELD(ExtraTrackPieces, {TRACK_SLOPE_VERTICAL, TRACK_TWIST, TRACK_CORKSCREW, TRACK_CURVE_VERTICAL, TRACK_QUARTER_LOOP, TRACK_SLOPE_LONG, TRACK_HALF_LOOP, TRACK_HALF_LOOP_MEDIUM, TRACK_HALF_LOOP_LARGE,TRACK_CORKSCREW_LARGE}),
```
Comment (Generalcamo) [inline src/openrct2/ride/coaster/meta/LoopingRollerCoaster.h:23]: If we're including the large corkscrew into the extra track pieces, and we already have the twist available here, should probably include the barrel roll too.

**Review by Generalcamo — COMMENTED** [review]
Some sprites for booster pieces are included for the LIM, but I notice no changes from the default  straight track pieces. Was this an accidental inclusion?

File: src/openrct2/ride/coaster/meta/LoopingRollerCoaster.h (hunk 2 above)
Comment (X123M3-256) [reply src/openrct2/ride/coaster/meta/LoopingRollerCoaster.h]: Yeah, this was a mistake - the zero G rolls were also missing. It has been fixed now.

**Comment by X123M3-256** [comment]
> Some sprites for booster pieces are included for the LIM, but I notice no changes from the default straight track pieces. Was this an accidental inclusion?

Yes, the original plan was to add the booster sprites as part of this PR so I added some placeholder sprites. I've now decided to do that in a separate PR as I don't know how long it'll take to come up with a suitable booster sprite, and I'm thinking of adding vertical boosters as a new track piece.

I've now removed the unused booster sprites and associated sprite IDs from this branch.

File: src/openrct2/ride/coaster/meta/LIMLaunchedRollerCoaster.h (hunk 1 above)
Comment (Gymnasiast) [reply src/openrct2/ride/coaster/meta/LIMLaunchedRollerCoaster.h:22]: Please leave these things to us.

**Comment by Gymnasiast** [comment]
Did you have a specific reason to enable ”Large steep to flat” pieces on the Looping RC by default?

**Comment by X123M3-256** [comment]
> Did you have a specific reason to enable ”Large steep to flat” pieces on the Looping RC by default?

Is there any reason not to? It doesn't require any sprites that the looping trains don't have and it's a useful track piece to have.

**Comment by Gymnasiast** [comment]
> > Did you have a specific reason to enable ”Large steep to flat” pieces on the Looping RC by default?
> 
> Is there any reason not to? It doesn't require any sprites that the looping trains don't have and it's a useful track piece to have.

To me, it doesn’t seem to fit the rest of the coaster type, considering it’s usually compact, twisty designs.

**Comment by spacek531** [comment]
> > > Did you have a specific reason to enable ”Large steep to flat” pieces on the Looping RC by default?
> > 
> > 
> > Is there any reason not to? It doesn't require any sprites that the looping trains don't have and it's a useful track piece to have.
> 
> To me, it doesn’t seem to fit the rest of the coaster type, considering it’s usually compact, twisty designs.

This decision should be made by the user, in my opinion.

**Comment by Gymnasiast** [comment]
> > > > Did you have a specific reason to enable ”Large steep to flat” pieces on the Looping RC by default?
> > > 
> > > 
> > > Is there any reason not to? It doesn't require any sprites that the looping trains don't have and it's a useful track piece to have.
> > 
> > 
> > To me, it doesn’t seem to fit the rest of the coaster type, considering it’s usually compact, twisty designs.
> 
> This decision should be made by the user, in my opinion.

We have hidden more track elements behind this flag that vehicles are technically capable of, but that would be unrealistic. This includes stuff like the holding brake on the Twister RC.

IMO, the user already has a way to decide and that is to turn on the cheat.

Incidentally, this is a minor thing and we can very easily revisit the decision at any time.

**Comment by X123M3-256** [comment]
> We have hidden more track elements behind this flag that vehicles are technically capable of, but that would be unrealistic. 

This isn't a piece that would be unrealistic. It's not really unrealistic for any ride type except perhaps the wild mouse (and of course those that don't have steep slopes at all) - the standard size slope transitions are very small. 

Arguably, the medium and large loops are realistic as well, but I left those disabled as I don't think they really fit.

**Comment by ZeeMajora** [comment]
I think the large steep to flat pieces are fine on the looping RC and don't really look that out of place to have them available by default, you can already kind of fake them with the normal incline pieces this just looks much nicer.

Screenshot_select-area_20221002181251

**Comment by Gymnasiast** [comment]
Fine, let’s leave it enabled by default then.

**Review by Gymnasiast — APPROVED** [review]

**Comment by Gymnasiast** [comment]
I did a quick test and it seems to work nicely.  I have asked @rik-smeets to test it as well and make sure there is no z-fighting, clipping or anything like that.

**Review by rik-smeets — CHANGES_REQUESTED** [review]
I noticed some issues, could you try to fix these? You can use this park to reproduce:
[LIM new track pieces.zip](https://github.com/OpenRCT2/OpenRCT2/files/9708469/LIM.new.track.pieces.zip)

1. Clipping on large flat to steep
Clipping large flat to steep

2. Zero G rolls
Trains passing through the zero G rolls don't look exactly right, clipping through the tracks. In case of the large zero G roll, also some clipping through the tree.
Large Zero G
Small Zero G

3. Disconnected barrel roll on flat land
image

**Comment by karst** [comment]
> Zero G rolls
Trains passing through the zero G rolls don't look exactly right, clipping through the tracks. In case of the large zero G roll, also some clipping through the tree.

Could you test this with a train that supports zero-g rolls?

**Comment by karst** [comment]
image


The first one is also a train related bug, I imagine that'd have to be fixed in the objects directory.

**Comment by X123M3-256** [comment]
@rik-smeets 

> Clipping on large flat to steep

I can't figure out the cause of this issue, but I don't think it's related to the track piece -  if you put the LIM trains on giga or twister track, they still clip through the tree. In fact, they still clip with the track made invisible. But if you put the giga or twister trains on the LIM track, they don't clip. The bounding boxes for this track piece were copied from the giga coaster so I wouldn't expect different behaviour.

The fact that the other train types don't have a problem suggests it's an issue with the vehicle, but unlike other vehicle related glitches, the bounding boxes look reasonable so I'm not sure why this happens. In any case, I don't think any changes to the track piece will prevent it.

> Trains passing through the zero G rolls don't look exactly right, clipping through the tracks. In case of the large zero G roll, also some clipping through the tree.

The clipping through the tree is the same issue as on the large steep to flat. I'm not sure what you mean about the trains ... of course they're not going to look right when you use a train that can't do zero G rolls, did you test with the hybrid or single rail train?

> Disconnected barrel roll on flat land

That one definitely is a problem with the track piece, I'll push a fix in a minute.

**Comment by Gymnasiast** [comment]
@rik-smeets Could you repeat the clipping test using the Looping RC train?

**Comment by X123M3-256** [comment]
The barrel roll issue should be fixed now

**Comment by rik-smeets** [comment]
@Gymnasiast The Looping RC trains don't clip, so as @karst said it seems to be an issue with the vehicles. I'll create a separate issue for it.

@X123M3-256 You're right, as @Gymnasiast and @karst said: I was mistakenly using the wrong train type. Sorry for the confusion.

I'll retest the barrel roll with the Windows build as soon as it's completed, thanks for the quick action!

**Comment by rik-smeets** [comment]
The issue is fixed, but the barrel roll (left) doesn't properly connect to the adjacent piece of track:
image
image


This doesn't happen with the barrel roll (right). Could you fix this too @X123M3-256 ?

**Comment by SpartanFrederic104** [comment]
Does it happen only on this specific angle?

**Comment by rik-smeets** [comment]
> Does it happen only on this specific angle?

Yes, I only noticed it at that angle.

**Comment by X123M3-256** [comment]
> This doesn't happen with the barrel roll (right). Could you fix this too X123M3-256 ?

I already went through every sprite and edited them by hand to try and smooth out the transitions as much as I could but I'm wondering if I missed that one, because it does look like it could be made a little smoother than it is. 

I'll take another look at it, but it's impossible to make every piece line up because the existing pieces don't all match up with each other. For example, the flat to gentle slope pieces don't line up with the straight gentle slope pieces, so the half loops won't always line up if you place a gentle slope piece before (I decided that the flat to gentle into a half loop would be the more common situation). Because of this, there's quite a few combinations of vanilla pieces that don't connect smoothly too.

**Comment by X123M3-256** [comment]
OK, I've touched up that sprite a bit, that's probably the best I can do with it. It looks better than before, I think.

**Comment by rik-smeets** [comment]
> OK, I've touched up that sprite a bit, that's probably the best I can do with it. It looks better than before, I think.

It does look better, thanks!

**Review by rik-smeets — APPROVED** [review]

**Pull request merged**